enhancements/machine-config/pin-and-pre-load-images.md

```diff
@@ -0,0 +1,424 @@
+see-also:
+- https://github.com/openshift/enhancements/pull/1432
+- https://github.com/openshift/api/pull/1609
+replaces: []
+superseded-by: []
+---
+
+# Pin and pre-load images
+
+## Summary
+
+Provide an mechanism to pin and pre-load container images.
+
+## Motivation
+
+Slow and/or unreliable connections to the image registry servers interfere with
+operations that require pulling images. For example, an upgrade may require
+pulling more than one hundred images. Failures to pull those images cause
+retries that interfere with the upgrade process and may eventually make it
+fail. One way to improve that is to pull the images in advance, before they are
+actually needed, and ensure that they aren't removed. Doing that provides a
+more consistent upgrade time in those environments. That is important when
+scheduling upgrades into maintenance windows, even if the upgrade might not
+otherwise fail.
+
+### User Stories
+
+#### Pre-load and pin upgrade images
+
+As the administrator of a cluster that has a low bandwidth and/or unreliable
+connection to an image registry server I want to pin and pre-load all the
+images required for the upgrade in advance, so that when I decide to actually
+perform the upgrade there will be no need to contact that slow and/or
+unreliable registry server and the upgrade will successfully complete in a
+predictable time.
+
+#### Pre-load and pin application images
+
+As the administrator of a cluster that has a low bandwidth and/or unreliable
+connection to an image registry server I want to pin and pre-load the images
+required by my application in advance, so that when I decide to actually deploy
+it there will be no need to contact that slow and/or unreliable registry server
+and my application will successfully deploy in a predictable time.
+
+### Goals
+
+Provide a mechanism that cluster administrators can use to pin and pre-load
+container images.
+
+### Non-Goals
+
+We wish to use the mechanism described in this enhancement to orchestrate
+upgrades without a registry server. But that orchestration is not a goal
+of this enhancement; it will be part of a separate enhancement based on
+parts of https://github.com/openshift/enhancements/pull/1432.
+
+## Proposal
+
+### Workflow Description
+
+1. The administrator of a cluster uses the new `PinnedImageSet` custom resource
+to request that a set of container images are pinned and pre-loaded:
+
+    ```yaml
+    apiVersion: machineconfiguration.openshift.io/v1alpha1
+    kind: PinnedImageSet
+    metadata:
+      name: my-pinned-images
+    spec:
+      nodeSelector:
+        matchLabels:
+          node-role.kubernetes.io/control-plane: ""
+      pinnedImages:
+      - quay.io/openshift-release-dev/ocp-release@sha256:...
+      - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+      - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+      ...
+    ```
+
+1. The machine config operators ensures that all the images are pinned and
```

Comment:
this is one final thing I am not clear on: who is pulling the images?

Reply:
I feel like we should determine that before we move forward, though I could be convinced otherwise

Reply:
The suggestion is to have a new `PinnedImageSetController` sub-controller in the _machine-config-controller_. I will try to clarify that in this sentence. It is also explained in the implementation details section:

> The _machine-config-controller_ will grow a new `PinnedImageSetController` sub
> controller that will watch the `PinnedImageSet` custom resources. When one of
> those is created, updated or deleted the controller will start a daemon set that
> will do the following in each node of the cluster:
> 
> 1. Verify that there is enough disk space available in the machine to pull the
>    images.
> 
> 1. Create, modify or delete the CRI-O pinning configuration file.
>
> 1. Reload the CRI-O configuration, with the equivalent of `systemctl reload crio`.
> 
>     Note that currently CRI-O doesn't recalculate the set of pinned images on
>    reload, support for that will need to be added.
> 
>  1. Use the CRI-O gRPC API to run the equivalent of `crictl pull` for each of the
>  images.

This sub-controller can't pull the images directly because that requires talking to the gRPC API of CRI-O, and that is only available via `localhost` in each node. It is for that reason that I am suggesting to add a daemon set to do it. It could also be part of the _machine-config-daemon_ instead of a new daemon set, but I am not familiar enough with that code to say what is better. I also tried to explain that in the implementation details section.

Ultimately the images will be pulled by the CRI-O running in each node, think of it as going to each node and running `crictl pull ...` for each image.

Reply:
I see this similar to how MCO applies kubeletconfig and containerruntime config on node. kubeletconfig and containerruntime controller processes respective CRD applied to the cluster which in results renders a new MachineConfig. Node sub-controller applies desired config annotation on node to signal machine-config-deamon (MCD running on each node ) to apply the change. In similar way, PinnedImageSetController would process the applied PinnedImageSetController CRD and a new rendered MachineConifg gets generated. MCD learns to apply the change by pulling those images using crictl pull and whatever needs to be done locally on the node. Alternatively, we can introduce a new daemon if because of any unknown limitation MCD can't handle it. Note that introducing a new daemon would still require some sort of co-ordination so that they don't go sideways like MCD started drain/reboot due a change and at the same time new daemon to pulling the images)